### PR TITLE
Be able to add location in engagement report

### PR DESCRIPTION
**Pull request by pacodelcastillolopez**
Some customers want regulars users to be able to add locations from within the report form if they do not exist. This featureis configurable through the dictionary.
If activated, when the user types a location name in the report form and no matches are found, the "no results found" message will also contain a link to open the new location form within the pop-up. The user can then create a new location which gets automatically assigned to the report.

Closes [AB#996](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/996)

#### User changes
- They can create locations within the report form if allowed in the dictionary.

#### Superuser changes
- They can always create locations within the report form.

#### Admin changes
- None.

#### System admin changes
- [x] anet.yml or anet-dictionary.yml needs change
  ```
  regularUsersCanCreateLocations: false
  […]
  fields:
    […]
    location:
      […]
      superuserTypeOptions: [POINT_LOCATION]
      regularuserTypeOptions: [POINT_LOCATION]
      […]
  ```
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [x] Added and/or updated unit tests
  - [ ] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [x] Opened debt issues for anything not resolved here

### CODE REVIEW

File: client/src/pages/reports/Form.js [hunk 1]
```diff
@@ -572,6 +575,26 @@ const ReportForm = ({
...
+                                setFieldValue("location", value, true)
+                                setDoReset(true)
+                                toast.success("The location has been saved")
+                              }}
+                              afterCancelActions={() => {
+                                setDoReset(true)
+                              }}
+                            />
+                          ))) ||
+                        null
```
Comment (gjvoosten) [inline client/src/pages/reports/Form.js:596]: For readability, I'd suggest to extract this as a component, e.g.:
```javascript
const CreateNewLocation = ({
  canCreateNewLocation,
  searchText,
  setFieldTouched,
  setFieldValue,
  setDoReset
}) => {
  if (!canCreateNewLocation) {
    return null
  }

  return (
    <LocationForm
      initialValues={new Location({ name: searchText })}
      title="Create New Location"
      afterSaveActions={value => {
        // validation will be done by setFieldValue
        setFieldTouched("location", true, false) // onBlur doesn't work when selecting an option
        setFieldValue("location", value, true)
        setDoReset(true)
        toast.success("The location has been saved")
      }}
      afterCancelActions={() => {
        setDoReset(true)
      }}
    />
  )
}

CreateNewLocation.propTypes = {
  canCreateNewLocation: PropTypes.bool,
  searchText: PropTypes.string,
  setFieldTouched: PropTypes.func.isRequired,
  setFieldValue: PropTypes.func.isRequired,
  setDoReset: PropTypes.func.isRequired
}
```
then call it like this:
```javascript
                      createEntityComponent={(searchTerms, setDoReset) => (
                        <CreateNewLocation
                          canCreateNewLocation={canCreateLocation}
                          searchTerms={searchTerms}
                          setFieldTouched={setFieldTouched}
                          setFieldValue={setFieldValue}
                          setDoReset={setDoReset}
                        />
                      )}
```

File: client/src/pages/locations/Form.js [hunk 2]
```diff
@@ -56,16 +56,25 @@ const LOCATION_TYPES_ADMIN = [
...
+  afterCancelActions
+}) => {
   const { currentUser } = useContext(AppContext)
   const navigate = useNavigate()
   const [error, setError] = useState(null)
   const [attachmentList, setAttachmentList] = useState(
     initialValues?.attachments
   )
   const [showSimilarLocations, setShowSimilarLocations] = useState(false)
+  const regularUsersCanCreateLocations = Settings.regularUsersCanCreateLocations
```
Comment (gjvoosten) [inline client/src/pages/locations/Form.js:74]: Note that regular users can't create a location as the changes are incomplete: the location type is required but regular users can't select a location type in the dropdown. Additional changes to this Form are needed!

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: anet-dictionary.yml [hunk 3]
```diff
@@ -3,6 +3,7 @@ CONNECTION_ERROR_MSG: Connection to the server is lost
 VERSION_CHANGED_MSG: "There is a new version of ANET, click here to load the new version (CAUTION: save any unsaved changes first)"
 SUPPORT_EMAIL_ADDR: support@example.com
 
+regularUsersCanCreateLocations: false
```
Comment (gjvoosten) [inline anet-dictionary.yml:6]: We should probably set this to `true` in our development dictionaries so we use the functionality ourselves.
And then also change the server-side test (i.e. reverse the conditions).

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 4]
```diff
@@ -332,7 +337,36 @@ const AdvancedSelect = ({
...
+                                  <div>
+                                    No results found.
+                                    <span
+                                      id="createEntityLink"
+                                      className="asLink"
+                                      onClick={() =>
+                                        setShowCreateEntityComponent(true)}
+                                    >
+                                      Click here to create a new {fieldName}
+                                    </span>
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:360]: Prefer to use a button for this.

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 5]
```diff
@@ -332,7 +337,36 @@ const AdvancedSelect = ({
                         renderRow={overlayRenderRow}
                         isLoading={isLoading}
                         loaderMessage={
-                          <div style={{ width: "300px" }}>No results found</div>
+                          (createEntityComponent && (
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:340]: I would either show result of `createEntityComponent`, or the `OverlayTable` and the `UltimatePagination`, as the latter doesn't make sense at all when creating a new entity.

File: client/src/pages/locations/Form.js (hunk 2 above)
Comment (pacodelcastillolopez) [reply client/src/pages/locations/Form.js:74]: :( I was sure I had tested this...

File: ANET.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+./gradlew dockerStartDB
```
Comment (gjvoosten) [inline ANET.sh:1]: Please remove this file from this PR.

File: ANET_test.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+./gradlew -PtestEnv dockerStartDB
```
Comment (gjvoosten) [inline ANET_test.sh:1]: Please remove this file from this PR.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/advancedSelectWidget/AdvancedSelect.js [hunk 8]
```diff
@@ -460,6 +486,7 @@ const AdvancedSelect = ({
     setPageNum(results && filterResults ? filterResults.pageNum : 0)
     setIsLoading(shouldFetchResults)
     setFetchType(shouldFetchResults ? FETCH_TYPE.NORMAL : FETCH_TYPE.NONE)
+    setShowCreateEntityComponent(false)
```
Comment (gjvoosten) [inline client/src/components/advancedSelectWidget/AdvancedSelect.js:489]: This line is unnecessary; `showCreateEntityComponent` will always already be `false` here.

File: client/src/components/GeoLocation.js [hunk 9]
```diff
@@ -348,7 +348,7 @@ const CoordinateActionButtons = ({
   setLocationFormat
 }) => {
   return (
-    <Col sm={3} style={{ padding: "0 8px" }}>
+    <Col sm={4} style={{ padding: "0 8px" }}>
```
Comment (gjvoosten) [inline client/src/components/GeoLocation.js:351]: This change is unnecessary.

**Review by gjvoosten — APPROVED** [review]

**Pull request merged**